app/src/main/java/com/epfl/drawyourpath/userProfile/cache/UserModelCached.kt

```diff
@@ -274,16 +306,44 @@ class UserModelCached(application: Application) : AndroidViewModel(application)
         future.thenApplyAsync {
             database.addDailyGoal(currentUserID!!, DailyGoal(it.first))
             it.second
-        } /*.thenComposeAsync { TODO uncomment when function is created
-            database.addMilestones(currentUserID!!, it.map {entity -> Milestone(entity)})
-        }*/.thenComposeAsync {
+        }.thenComposeAsync {
+            addListMilestones(it)
+        }.thenComposeAsync {
             database.addRunToHistory(currentUserID!!, run)
         }.thenApplyAsync {
             runCache.runSynced(currentUserID!!, run.getStartTime())
         }
         return future.thenApply {}
     }
 
+    /**
+     * Helper function to add all the milestones in the database
+     * @param milestones list of milestones to add in the database
+     * @return a completable future that indicate if the the milesstones were correctly added
+     */
+    private fun addListMilestones(milestones: List<MilestoneEntity>): CompletableFuture<Unit> {
+        var future = CompletableFuture<Unit>()
+        if (milestones.isEmpty()) {
+            future.complete(Unit)
+        } else {
+            milestones.forEachIndexed { index, entity ->
+                val newFuture = database.addMilestone(
+                    milestone = MilestoneEnum.valueOf(
+                        Utils.getALL_CAPSFromString(entity.milestone),
+                    ),
+                    date = LocalDate.ofEpochDay(entity.date),
+                    userId = entity.userId,
+                )
+                if (index == 0) {
+                    future = newFuture
+                } else {
+                    future.thenApply { newFuture }
+                }
+            }
+        }
+        return future
+    }

```

Comment:
This doesn't propagate the error if one of the future (except the first one) throw an exception.
Here is an alternative:
```kotlin
private fun addListMilestones(milestones: List<MilestoneEntity>): CompletableFuture<Void> {
        return CompletableFuture.allOf(
            *milestones.map {
                database.addMilestone(
                    MilestoneEnum.valueOf(Utils.getALL_CAPSFromString(it.milestone)),
                    LocalDate.ofEpochDay(it.date),
                    it.userId
                )
            }.toTypedArray()
        )
    }
```